### PR TITLE
[diffusion] model: simplifie VAE cache logic for  Wan

### PR DESCRIPTION
**Pull request by triple-mu**
## Motivation

Simplified VAE cache logic for  Wan

This PR simplifies the VAE cache logic used in  Wan.

For regular CausalConv3d, the cache is maintained by tracking and updating the temporal state via self.prev_cache.

Resample follows a different caching behavior, so a dedicated implementation is added to handle this case.

With this change, complex state management is no longer required.

However, it is still unclear whether this approach works correctly with parallel encoding or decoding, and this may require further validation.

## Checklist

- [✓] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [☓] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [☓] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [☓] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [✓] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).
- [☓] Work with maintainers to merge your PR. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process)

### CODE REVIEW

File: python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py [hunk 1]
```diff
@@ -1184,118 +999,89 @@ def __init__(
...
-                        first_chunk.set(False)
-                        out_ = self.decoder(x[:, :, i : i + 1, :, :])
-                        out = torch.cat([out, out_], 2)
-
+            out = []
+            for i in range(iter_):
+                if i == 0:
+                    out_ = self.decoder(x[:, :, i : i + 1, :, :])
+                else:
+                    out_ = self.decoder(x[:, :, i : i + 1, :, :])
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py]: DONE

File: python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py [hunk 2]
```diff
@@ -188,29 +164,65 @@ def __init__(
...
+        x_with_cache = (
+            x_with_cache.to(self.weight.dtype)
+            if current_platform.is_mps()
+            else x_with_cache
         )  # casting needed for mps since amp isn't supported
-        return super().forward(x)
+        x = super().forward(x_with_cache)
+        x = rearrange(x, "b (r c) t h w -> b c (t r) h w", r=2)
+        self.prev_cache = x_with_cache.narrow(2, t, self.pad_t)
+        return x
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py]: DONE

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 3]
```diff
@@ -962,16 +830,16 @@ def _decode(self, z: torch.Tensor, return_dict: bool = True):
...
+        out = []
         for i in range(num_frame):
-            self._conv_idx = [0]
             if i == 0:
-                out = self.decoder(x[:, :, i: i + 1, :, :], feat_cache=self._feat_map, feat_idx=self._conv_idx)
+                out_ = self.decoder(x[:, :, i: i + 1, :, :])
             else:
-                out_ = self.decoder(x[:, :, i: i + 1, :, :], feat_cache=self._feat_map, feat_idx=self._conv_idx)
-                out = torch.cat([out, out_], 2)
+                out_ = self.decoder(x[:, :, i: i + 1, :, :])
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: DONE

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 4]
```diff
@@ -51,26 +53,65 @@ def __init__(
...
+        x_with_cache = torch.cat([self.prev_cache, x], dim=2)
+        x_with_cache = (
+            x_with_cache.to(self.weight.dtype)
+            if current_platform.is_mps()
+            else x_with_cache
+        )  # casting needed for mps since amp isn't supported
+        x = super().forward(x_with_cache)
+        x = rearrange(x, "b (r c) t h w -> b c (t r) h w", r=2)
+        self.prev_cache = x_with_cache.narrow(2, t, self.pad_t)
+        return x
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py]: DONE

**Review by mickqian — COMMENTED** [review]
Could you verify locally that the accuracy has not dropped?

**Comment by triple-Mu** [comment]
> hi, please do as described in: https://github.com/sgl-project/sglang/blob/main/python/sglang/multimodal_gen/docs/contributing.md#2-performance-reporting

benchmark command:

```bash
#!/bin/bash

prompt="A cat and a dog baking a cake together in a kitchen. The cat is carefully measuring flour, while the dog is stirring the batter with a wooden spoon. The kitchen is cozy, with sunlight streaming through the window."
negative_prompt="Bright tones, overexposed, static, blurred details, subtitles, style, works, paintings, images, static, overall gray, worst quality, low quality, JPEG compression residue, ugly, incomplete, extra fingers, poorly drawn hands, poorly drawn faces, deformed, disfigured, misshapen limbs, fused fingers, still picture, messy background, three legs, many people in the background, walking backwards"

sglang generate \
    --model-path Wan2.1-T2V-14B-Diffusers \
    --prompt "${prompt}" \
    --negative-prompt "${negative_prompt}" \
    --num-gpus 8 \
    --ulysses-degree 8 \
    --dit-cpu-offload false \
    --use-fsdp-inference false \
    --text-encoder-cpu-offload false \
    --image-encoder-cpu-offload false \
    --vae-cpu-offload false \
    --num-frames 81 \
    --height 1280 \
    --width 720 \
    --fps 15 \
    --num-inference-steps 50 \
    --guidance-scale 5.0 \
    --save-output \
    --perf-dump-path baseline.json

sglang generate \
    --model-path Wan2.1-T2V-14B-Diffusers \
    --prompt "${prompt}" \
    --negative-prompt "${negative_prompt}" \
    --num-gpus 8 \
    --ulysses-degree 8 \
    --dit-cpu-offload false \
    --use-fsdp-inference false \
    --text-encoder-cpu-offload false \
    --image-encoder-cpu-offload false \
    --vae-cpu-offload false \
    --num-frames 81 \
    --height 1280 \
    --width 720 \
    --fps 15 \
    --num-inference-steps 50 \
    --guidance-scale 5.0 \
    --save-output \
    --perf-dump-path my_pr.json

python python/sglang/multimodal_gen/benchmarks/compare_perf.py baseline.json my_pr.json
```

result:

**Review by mickqian — CHANGES_REQUESTED** [review]
CI shows potential performance regression

File: python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py [hunk 5]
```diff
@@ -1290,12 +1064,11 @@ def decode(self, z: torch.Tensor) -> torch.Tensor:
         return out
 
     def _decode(self, z: torch.Tensor, first_frame=False) -> torch.Tensor:
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py]: DONE

File: python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py [hunk 6]
```diff
@@ -1290,12 +1064,11 @@ def decode(self, z: torch.Tensor) -> torch.Tensor:
...
+        self.clear_cache()
         x = self.post_quant_conv(z)
-        with forward_context(first_frame_arg=first_frame):
-            out = self.decoder(x)
-
+        out = self.decoder(x)
         out = torch.clamp(out, min=-1.0, max=1.0)
-
+        self.clear_cache()
         return out
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py]: Your understanding is incorrect. In VAE chunking, there is no dependency between different chunks, so the cache does not need to be preserved across chunks.

File: python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py [hunk 7]
```diff
@@ -888,41 +752,33 @@ def disable_slicing(self) -> None:
...
-
-    def _encode(self, x: torch.Tensor):
+    def clear_cache(self) -> None:
+        for m in self.encoder.modules():
+            if isinstance(m, CausalConv3d):
+                m.clear_cache()
+        for m in self.decoder.modules():
+            if isinstance(m, CausalConv3d):
+                m.clear_cache()
+        torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/autoencoder_kl_qwenimage.py:766]: VAE decoding and encoding lead to significant GPU memory usage, so it is necessary to call empty_cache.

File: python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py [hunk 8]
```diff
@@ -1184,100 +985,73 @@ def __init__(
...
         return enc
 
     def spatial_tiled_encode(self, x: torch.Tensor) -> torch.Tensor:
-        first_frame = x[:, :, 0, :, :].unsqueeze(2)
-        first_frame = self._encode(first_frame, first_frame=True)
-
         enc = ParallelTiledVAE.spatial_tiled_encode(self, x)
-        enc = enc[:, :, 1:]
-        enc = torch.cat([first_frame, enc], dim=2)
         return enc
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py:1040]: The computation of the first frame is mainly for creating the cache, because the cache shape may change later. However, the maximum temporal length of the cache does not change, so a separate first-frame operation is not necessary.

File: python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py [hunk 9]
```diff
@@ -1184,100 +985,73 @@ def __init__(
...
+
+        # cache encoder
         if self.config.load_encoder:
-            self._enc_conv_num = _count_conv3d(self.encoder)
-            self._enc_conv_idx = 0
-            self._enc_feat_map = [None] * self._enc_conv_num
+            for m in self.encoder.modules():
+                if isinstance(m, CausalConv3d):
+                    m.clear_cache()
+        torch.cuda.empty_cache()
```
[the remark this answers is not in the document]
Comment (triple-mu) [reply python/sglang/multimodal_gen/runtime/models/vaes/wanvae.py]: VAE decoding and encoding lead to significant GPU memory usage, so it is necessary to call empty_cache.
